Clims 257 extendible projects

johandahlberg opened the pull request:
This adds the concept of projects to clims. There are still some things to do - e.g. writing tests for the project service, but since this also contains some general refactoring of how the other services work with regards to wrapping objects, so I thought it would be good idea to get this into the develop branch.

Part of this looks a bit odd, since I had to move some code around in the service class to get the interpreter to have things defined in the correct order. But all that really happened there is that I moved the wrapper methods to their own mixin.

This also fixes two minor bugs:
 - It removes the submodule that I accidentally added in previous PR
 - Fixes the create example data script to match up with the changes to the API.

johandahlberg committed:
Removed accidentally added submodule

johandahlberg committed:
Add new clims.Project model and related entities

Commit:
Fix createexampledata import

Commit:
Move wrapping code to mixin

Commit:
Associate substance with clims.Project

withrocks commented on src/clims/models/project.py:
```diff
@@ -0,0 +1,42 @@
+    name = models.TextField()
+    organization = FlexibleForeignKey('sentry.Organization', related_name='%(class)s_organization')
+
+    # TODO What other properties do we want to be native to the projects?
```
I agree. I guess we can just drag things up from the properties as we progress, in particular after we've created the project subclass in commonlims-snpseq.

johandahlberg replied:
:+1:

withrocks commented on src/clims/models/substance.py:
```diff
@@ -26,9 +26,9 @@ def __init__(self, *args, **kwargs):
     name = models.TextField()
 
-    organization = FlexibleForeignKey('sentry.Organization')
+    organization = FlexibleForeignKey('sentry.Organization', related_name='%(class)s_organization')
```
Wouldn't the related_name just be `substances`? That would mean that if you have an Organization, you can query for the substances using this syntax:

```
org = Organization.get(name=lab)
org.substances.all()
```

As this is now (if I'm not missing something) it would be:

```
org.substances_organization.all()
```

johandahlberg replied:
I think you are right. I'll fix this.

withrocks commented on src/clims/services/container.py:
```diff
@@ -267,3 +238,16 @@ def to_table():
         return "\n".join(map(six.text_type, rows))
+
+
+class ContainerService(WrapperMixin):
```
This is great. I believe we should also inherit from object, since we're still using Python 2:

https://stackoverflow.com/a/38140423/282024

johandahlberg replied:
Good catch!

withrocks commented on src/clims/services/project.py:
```diff
@@ -0,0 +1,49 @@
+        """
+        return self._app.projects.to_wrapper(model)
+
+    @property
```
I guess we would later also want to add these to a mixin.

johandahlberg replied:
Your thinking of the properties? Yes, I suppose so. Or be moved into the `ExtensibleBase`?

withrocks reviewed (commented):
Looks great! I wonder if you want to change that related_name before merging.

Commit:
Fix related names in project/substance models

Commit:
Have services inherit from object (py2 comp.)

Commit:
ContainerBase and ProjectBase non-registered

GitEdvard commented on tests/clims/models/test_project.py:
```diff
@@ -0,0 +1,57 @@
+    def test_can_create_project(self):
+        project = VampireFangStudyProject(name="project1", organization=self.organization)
+        project.save()
+        Project.objects.get(name=project.name)  # Raises DoesNotExist if it wasn't created
```
I'm concerned about using Project class directly in tests. I think the tests should show a reader how to use the classes, and the class Project should not be used by upper layers, it should be wrapped (and hidden) within ProjectBase.

GitEdvard commented on src/clims/services/wrapper.py:
```diff
@@ -0,0 +1,35 @@
+    `_archetype_base`.
+    """
+
+    _archetype_version = None
```
Perhaps it could be indicated that these are class variables and not instances of some class?

GitEdvard commented:
Now I have also reviewed this PR. Nice work! I like the wrapper mixin base class.

Commit:
Clarify wrapper variable names

Commit:
Make API of clims "object services" mixin

This brings the api parts of the services out into it's own mixin that
can be used to access clims objects. This should make it easier for us
to iterate on this API as it is used similarly (identically?) across all
service classes.

GitEdvard commented on src/clims/services/extensible_service_api.py:
```diff
@@ -0,0 +1,45 @@
+        if 'latest' not in get_args.keys():
+            get_args['latest'] = True
+
+        substance_model = self._archetype_version_class.objects.prefetch_related('properties').get(**get_args)
```
I think it should not be substance_model :)

johandahlberg replied:
Damn you copy-paste! Why don't you automatically re-name stuff? :wink:

GitEdvard commented on src/clims/services/extensible_service_api.py:
```diff
@@ -0,0 +1,45 @@
+        get_args = {}
+        for key, value in kwargs.items():
+            get_args['archetype__{}'.format(key)] = value
+        if 'latest' not in get_args.keys():
```
If I understand 'not in \<list\>' correctly, it should only get true for exact matches. All keys in get_args begins with 'archetype'

johandahlberg replied:
Good point. Fix this now.

GitEdvard commented on src/clims/services/extensible_service_api.py:
```diff
@@ -0,0 +1,45 @@
+    def get(self, **kwargs):
+        get_args = {}
+        for key, value in kwargs.items():
+            get_args['archetype__{}'.format(key)] = value
```
The word 'archetype__' is hard coded here. Shouldn't it be replaced by the different sub type names?

johandahlberg replied:
No, because they all have a related field called archetype, which is what we tap into here.

GitEdvard replied:
Ok, good.

GitEdvard commented on tests/clims/models/test_project.py:
```diff
@@ -16,7 +15,7 @@ def setUp(self):
         project = VampireFangStudyProject(name="project1", organization=self.organization)
         project.save()
-        Project.objects.get(name=project.name)  # Raises DoesNotExist if it wasn't created
+        self.app.projects.get(name=project.name)  # Raises DoesNotExist if it wasn't created
```
I would like it to return None if it doesn't exist!

johandahlberg replied:
I suggest we keep the API as it is for now. But later, that might be what we want.

GitEdvard replied:
ok

GitEdvard reviewed (commented):
Super nice to remove a lot of duplication! I got a few comments. Tell me when you want to merge.

johandahlberg committed:
Fixing setup latest insertion on get

johandahlberg committed:
Added missing migration
